umpost/um2netcdf.py

```diff
@@ -319,6 +321,63 @@ def fix_latlon_coords(cube, grid_type, dlat, dlon):
     fix_lon_coord_name(longitude_coordinate, grid_type, dlon)
 
 
+def get_default_fill_value(cube):
+    """
+    Get the default fill value for a cube's data's type.
+
+    Parameters
+    ----------
+    cube: Iris cube object.
+
+    Returns
+    -------
+    fill_value: numpy scalar with type matching cube.data
+    """
+    if cube.data.dtype.kind == 'f':
+        fill_value = DEFAULT_FILL_VAL_FLOAT
+
+    else:
+        fill_value = default_fillvals[
+            f"{cube.data.dtype.kind:s}{cube.data.dtype.itemsize:1d}"
+        ]
+
+    # TODO: We want to ensure the fill value matches the type of the
+    # cube's data. DEFAULT_FILL_VAL_FLOAT and netCDF4.default_fillvals
+    # use standard python floats and ints. Is there a cleaner
+    # way do do the following conversion?
+    return np.array([fill_value], dtype=cube.data.dtype)[0]
+
+
+def fix_fill_value(cube, custom_fill_value=None):
+    """
+    Set a cube's missing_value attribute.
+
+    Parameters
+    ----------
+    cube: Iris cube object (modified in place).
+    custom_fill_value: Fill value to use in place of defaults.
+    Type must match cube data's type.
+    """
+    if custom_fill_value is not None:
+        fill_value = custom_fill_value
+    else:
+        fill_value = get_default_fill_value(cube)
+
+    if type(fill_value) == cube.data.dtype:
```

Comment:
My linter (flake8) complains here, suggesting `do not compare types, for exact checks use "is" / "is not", for instance checks use "isinstance()"`

I've had a play around, however neither `is` or `isinstance` work as desired when using `cube.data.dtype` in the comparison. Would be keen to gather any suggestions!

Reply:
> ... neither `is` or `isinstance` work as desired when using `cube.data.dtype` in the comparison. Would be keen to gather any suggestions!

`is` will fail if the instances are not the same object.

The second aspect is `numpy` using internal types in place of Python's built in primitives:

```python
>>> type(cube.data.dtype)
numpy.dtypes.Float64DType
>>> type(1.0)
<class 'float'>
```
That affects comparison operations (more on that in another comment).

Reply:
Quick idea, do you want to try:
* set an IDE breakpoint at the check type
* running a fill val test in your IDE, in debug mode
* seeing what values both sides of `==` evaluate to?

Reply:
Ok sweet, I've had a go at that and during the tests we get:
```
type(fill_value)
<class 'numpy.float32'>
cube.data.dtype
dtype('float32')
.
type(fill_value)
<class 'numpy.float64'>
cube.data.dtype
dtype('float64')
.
type(fill_value)
<class 'numpy.complex64'>
cube.data.dtype
dtype('complex64')
.
type(fill_value)
<class 'numpy.int32'>
cube.data.dtype
dtype('int32')
.
type(fill_value)
<class 'numpy.int64'>
cube.data.dtype
dtype('int64')
.
type(fill_value)
<class 'numpy.float64'>
cube.data.dtype
dtype('float64')
.
# this is the one that's meant to fail
type(fill_value)
<class 'numpy.float32'>
cube.data.dtype
dtype('int32')
```

I just realised, after splitting it into two functions `get_default_fill_value()` and `fix_fill_value()`, we're essentially ensuring that the fill value provided to `fix_fill_value()` uses a numpy datatype, apart from when a `custom_fill_value` argument is supplied.

Would it be better to replace the if statement with `if fill_value.dtype == cube.data.dtype:`  and require that any `custom_fill_value` arguments have a numpy datatype?

Reply:
To clarify something, was there a requirement elsewhere to add a custom fill value?

Reply:
No external requirement, though it was included as an addition in the original issue #99 

If it's preferable we could remove the custom fill value completely?

Reply:
> No external requirement, though it was included as an addition in the original issue #99
> 
> If it's preferable we could remove the custom fill value completely?

Whoops, I think I threw the custom value thing into the issue without considering the details.

Let's simplify by raising a NotImplementedError if a custom value is provided & comment out the type checks in a comment. Working in the function space has been useful to uncover complexity!

Reply:
That sound's good. I've swapped it to raise a `NotImplementedError` when the `custom_fill_value` argument is used, and added in a comment noting that type comparisons of some sort should be added if custom fill values are eventually implemented.